devito/finite_differences/finite_difference.py

```diff
@@ -235,54 +173,60 @@ def generic_derivative(expr, dim, fd_order, deriv_order, symbolic=False,
         the resulting stencil.
     deriv_order : int
         Derivative order, e.g. 2 for a second-order derivative.
-    stagger : Side, optional
-        Shift of the finite-difference approximation.
+    matvec : Transpose, optional
+        Forward (matvec=direct) or transpose (matvec=transpose) mode of the
+        finite difference. Defaults to `direct`.
     x0 : dict, optional
         Origin of the finite-difference scheme as a map dim: origin_dim.
+    symbolic : bool, optional
+        Use default or custom coefficients (weights). Defaults to False.
 
     Returns
     -------
     expr-like
         ``deriv-order`` derivative of ``expr``.
     """
+    side = None
     # First order derivative with 2nd order FD is highly non-recommended so taking
     # first order fd that is a lot better
     if deriv_order == 1 and fd_order == 2 and not symbolic:
         fd_order = 1
-    # Stencil positions
-    indices, x0 = generate_indices(expr, dim, fd_order, x0=x0)
+
+    return make_derivative(expr, dim, fd_order, deriv_order, side, matvec, x0, symbolic)
+
+
+def make_derivative(expr, dim, fd_order, deriv_order, side, matvec, x0, symbolic):
+    # The stencil positions
+    indices, x0 = generate_indices(expr, dim, fd_order, side=side, x0=x0)
 
     # Finite difference weights from Taylor approximation with these positions
     if symbolic:
-        c = symbolic_weights(expr, deriv_order, indices, x0)
+        weights = symbolic_weights(expr, deriv_order, indices, x0)
     else:
-        c = numeric_weights(deriv_order, indices, x0)
+        weights = numeric_weights(deriv_order, indices, x0)
 
-    return indices_weights_to_fd(expr, dim, indices, c, matvec=matvec.val)
+    return indices_weights_to_fd(expr, dim, indices, weights, matvec=matvec.val)
 
 
-def indices_weights_to_fd(expr, dim, inds, weights, matvec=1):
+def indices_weights_to_fd(expr, dim, indices, weights, matvec=1):
     """Expression from lists of indices and weights."""
-    diff = dim.spacing
-
-    d0 = ([d for d in expr.dimensions if d.root is dim] or [dim])[0]
-
-    mapper = {dim: d0, diff: matvec*diff}
-
-    # Loop through weights
+    mapper = {dim.spacing: matvec*dim.spacing}
     terms = []
-    for i, c in zip(inds, weights):
-        # Apply replacements to indices. Needs to be sympified in case
-        # the indices are pure numbers
+    for i, c in zip(indices, weights):
+        # Transpose FD if necessary through `mapper`
         try:
             iloc = i.xreplace(mapper)
         except AttributeError:
+            # Pure number -> sympify
             iloc = sympify(i).xreplace(mapper)
+        # Shift index due to staggering, if any
+        iloc -= expr.indices_ref[dim] - dim
+
         # Enforce fixed precision FD coefficients to avoid variations in results
-        c = sympify(c).evalf(_PRECISION)
+        v = sympify(c).evalf(_PRECISION)
```

Comment:
why `c -> v` c was for "coefficient"

Reply:
because you should not override the iteration variables